File: mobile/shared/src/commonMain/kotlin/com/automatelinux/pt/util/Geo.kt
package com.automatelinux.pt.util

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sqrt

/**
 * Straight-line metres between two coordinates.
 *
 * Equirectangular rather than haversine: over the distances this app deals in — a bus and
 * the person waiting for it — the error is centimetres, and the result is only ever shown
 * rounded to a tenth of a kilometre.
 *
 * Shared because "how far apart are these two points" now has several callers, and the
 * alternative to one function is the same three lines copied with a different sign
 * convention in each.
 */
fun metersBetween(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
    val dLat = (lat1 - lat2) * 111_320.0
    val dLon = (lon1 - lon2) * 111_320.0 * cos(lat2 * PI / 180.0)
    return sqrt(dLat * dLat + dLon * dLon)
}

File: frontend/public-transportation/lib/geo.ts
/**
 * Straight-line metres between two coordinates.
 *
 * Equirectangular rather than haversine: at the scale this app deals in — a bus, a stop,
 * a person — the error is centimetres, and every result is shown rounded to a tenth of a
 * kilometre.
 *
 * One copy. This formula was written out three times in this repo (this file's two
 * callers plus the Kotlin client), which is how three places came to agree on the maths
 * and disagree on nothing useful. Kotlin keeps its own in `util/Geo.kt` because it cannot
 * import TypeScript; everything on this side imports from here.
 */
export function metersBetween(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const dLat = (lat1 - lat2) * 111_320
  const dLon = (lon1 - lon2) * 111_320 * Math.cos((lat2 * Math.PI) / 180)
  return Math.sqrt(dLat * dLat + dLon * dLon)
}
